accepted/wasmtime-rpc.md

```diff
@@ -0,0 +1,162 @@
+
+[wRPC] is chosen as the RPC protocol for this RFC. See [(draft) specification](https://github.com/bytecodealliance/wrpc/blob/main/SPEC.md) for lower-level details.
+
+## Light on dependencies
+
+[`wrpc-transport`](https://docs.rs/wrpc-transport/latest/wrpc_transport/) crate, which would be the integration point for Wasmtime is built on top of well-known Rust crates, which are already in Wasmtime dependency graph:
+
+- [`bytes`](https://docs.rs/bytes/latest/bytes/)
+- [`futures`](https://docs.rs/futures/latest/futures/)
+- [`tokio-stream`](https://docs.rs/tokio-stream/latest/tokio_stream/)
+- [`tokio-util`](https://docs.rs/tokio-util/latest/tokio_util/)
+- [`tokio`](https://docs.rs/tokio/latest/tokio/)
+
+The only external wRPC dependency not already used by Wasmtime is [`pin-project-lite`](https://docs.rs/pin-project-lite/latest/pin_project_lite/).
+Other than that, wRPC depends on a few small, purpose-built crates maintained by wRPC developers and extracted from wRPC codebase itself.
+
+## Bytecode Alliance hosted project
+
+[wRPC] is a Bytecode Alliance hosted project and so Bytecode Alliance has full control over the project's development.
+
+For example, if the `pin-project-lite` dependency mentioned above is an issue - that can be promptly changed.
+
+## WIT as lingua franca
+
+[wRPC] is built on top of WIT, which is already utilizied throughout Wasmtime codebase and, in fact, is embedded in each and every component, therefore making each component compatible with [wRPC] out of the box.
+There is no need for a custom mapping scheme and/or naming convention as one would need with a different RPC protocol (e.g. gRPC).
+
+## WebAssembly as the wire format
+
+[wRPC] relies on [component model value definition encoding], which in turn is based on Core Wasm spec and Canonical ABI.
+In fact, it can be thought of as a "packed" Canonical ABI, optimized for data size.
+
+Usage of a familiar encoding here provides many optimization opportunities on the runtime level providing for efficient serialization and deserialization steps.
+
+## Familiar user experience
+
+[wRPC] provides `wit-bindgen-wrpc` binary and Rust crate, which are built on top of `wit-bindgen`, which provides a familiar user experience.
+For example, `wit_bindgen_wrpc::generate!` macro supports most of the options supported by `wit_bindgen::generate!`.
+
+Plugin developers familiar with component model would not need to learn a new technology to develop a Wasmtime plugin, rather use largely the same development flow that is already used for developing components.
+
+## Multi-language support
+
+Other than Rust, [wRPC] currently supports Go, but support for other languages is planned.
+Language binding generators benefit greatly from prior art in `wit-bindgen`, since a lot of that logic is reused.
+
+## Transport-agnostic
+
+[wRPC] is transport-agnostic, with 4 supported transports at the time of writing:
+- TCP
+- Unix domain sockets
+- QUIC
```

Comment:
How is authentication handled? Would it require a CA certificate or are self-signed certificates allowed?

Reply:
Depends on the implementation, ultimately `wasmtime` CLI user would be the one to decide, one thing to note though is that in current implementation, SAN is (ab?)used to specify the function being called: https://github.com/bytecodealliance/wrpc/blob/628e6bf79416ca2671d0a3c79ac25949f9bf84d8/crates/transport-quic/src/lib.rs#L29-L42

Currently, self-signed or not, the server (plugin) needs to generate a cert with appropriate SANs.

This can easily be changed/made configurable to e.g. be specified in the stream header as done for e.g. TCP and UDS https://github.com/bytecodealliance/wrpc/blob/628e6bf79416ca2671d0a3c79ac25949f9bf84d8/crates/transport/src/frame/conn/client.rs#L33-L43
I feel like being able to sign served plugin capabilities may come in quite handy, but I don't have strong feelings about this